3-0-spring-framework/3-3-0-enable-string-trimming/src/main/java/com/bobocode/TrimmedAnnotationBeanPostProcessor.java

```diff
@@ -1,21 +1,106 @@
+
+    @Override
+    public Object postProcessBeforeInitialization(Object bean, String beanName) throws BeansException {
+        if (containsParametersAnnotatedWithTrimmed(bean)) {
+            beansToBeProxied.put(beanName, bean);
+            beanInterceptors.put(beanName, new MethodInterceptor[]{provideTrimmingInterceptor()});
+        }
+        return BeanPostProcessor.super.postProcessBeforeInitialization(bean, beanName);
+    }
+
+    @Override
+    public Object postProcessAfterInitialization(Object bean, String beanName) throws BeansException {
+        Object toProxy = beansToBeProxied.get(beanName);
+        if (toProxy != null) {
+            Enhancer enhancer = new Enhancer();
+            enhancer.setSuperclass(toProxy.getClass());
+            setInterceptorsIfNeeded(enhancer, bean, beanName);
+            return enhancer.create();
+        }
+        return bean;
+    }
+
+    private MethodInterceptor provideTrimmingInterceptor() {
+        return (beanInstance, method, args, methodProxy) ->
+                methodProxy.invokeSuper(beanInstance, processParameters(method, args));
+    }
+
+    private Object[] processParameters(Method method, Object[] args) {
+        Parameter[] parameters = method.getParameters();
+        for (int i = 0; i < parameters.length; i++) {
+            if (parameters[i].isAnnotationPresent(Trimmed.class)) {
+                args[i] = trimMarkedString(args[i]);
+            }
+        }
+        return args;
+    }
+
+    private Object trimMarkedString(Object arg) {
+        if (arg instanceof String string) {
+            return StringUtils.hasLength(string) ? string.trim() : arg;
+        }
+        return arg;
+    }
+
+    private boolean containsParametersAnnotatedWithTrimmed(Object bean) {
+        return Arrays.stream(bean.getClass().getDeclaredMethods()).flatMap(m -> Stream.of(m.getParameters()))
+                .anyMatch(p -> p.isAnnotationPresent(Trimmed.class));
+    }
+
+    private void setInterceptorsIfNeeded(Enhancer enhancer, Object bean, String beanName) {
+        MethodInterceptor[] methodInterceptors = beanInterceptors.get(beanName);
+        if (methodInterceptors.length != 0) {
+            Callback[] interceptors = findInterceptors(bean);
+            if (interceptors.length != 0) {
+                MethodInterceptor[] newInterceptors = Arrays.copyOf(methodInterceptors, methodInterceptors.length + 1);
+                newInterceptors[newInterceptors.length - 1] = provideTrimmingInterceptor();
+                enhancer.setCallbacks(newInterceptors);
+            }
+            enhancer.setCallbacks(methodInterceptors);
+        }
+    }
+
+    @SneakyThrows
+    private MethodInterceptor[] findInterceptors(Object bean) {
+        var method = Arrays.stream(bean.getClass().getDeclaredMethods()).filter(m -> m.getName().equals("getCallbacks"))
+                .findFirst();
+        if (method.isPresent()) {
+            return (MethodInterceptor[]) method.get().invoke(bean);
```

Comment:
@Stanislav-Zabramnyi it looks scary! 😳

Reply:
agree. But It`s reflection))  

I will think how to do it in more elegant way.